js/packages/common/src/contexts/connection.tsx

```diff
@@ -13,171 +19,179 @@ import {
+    label: 'mainnet-beta (Serum)',
+    url: 'https://solana-api.projectserum.com/',
+    chainId: ChainId.MainnetBeta,
   },
   {
-    name: 'testnet' as ENV,
-    endpoint: clusterApiUrl('testnet'),
-    ChainId: ChainId.Testnet,
+    name: 'testnet',
+    label: 'testnet',
+    url: clusterApiUrl('testnet'),
+    chainId: ChainId.Testnet,
   },
   {
-    name: 'devnet' as ENV,
-    endpoint: clusterApiUrl('devnet'),
-    ChainId: ChainId.Devnet,
+    name: 'devnet',
+    label: 'devnet',
+    url: clusterApiUrl('devnet'),
+    chainId: ChainId.Devnet,
   },
 ];
 
-const DEFAULT = ENDPOINTS[0].endpoint;
+const DEFAULT_ENDPOINT = ENDPOINTS[0];
 
 interface ConnectionConfig {
   connection: Connection;
-  endpoint: string;
-  env: ENV;
-  setEndpoint: (val: string) => void;
-  tokens: TokenInfo[];
-  tokenMap: Map<string, TokenInfo>;
+  endpoint: Endpoint;
+  tokens: Map<string, TokenInfo>;
 }
 
 const ConnectionContext = React.createContext<ConnectionConfig>({
-  endpoint: DEFAULT,
-  setEndpoint: () => { },
-  connection: new Connection(DEFAULT, 'recent'),
-  env: ENDPOINTS[0].name,
-  tokens: [],
-  tokenMap: new Map<string, TokenInfo>(),
+  connection: new Connection(DEFAULT_ENDPOINT.url, 'recent'),
+  endpoint: DEFAULT_ENDPOINT,
+  tokens: new Map(),
 });
 
-export function ConnectionProvider({ children = undefined as any }) {
+export function ConnectionProvider({ children }: { children: any }) {
   const searchParams = useQuerySearch();
-  const network = searchParams.get('network');
-  const queryEndpoint =
-    network && ENDPOINTS.find(({ name }) => name.startsWith(network))?.endpoint;
+  const [networkStorage, setNetworkStorage] =
+    useLocalStorageState<ENDPOINT_NAME>('network', DEFAULT_ENDPOINT.name);
+  const networkParam = searchParams.get('network');
+
+  let maybeEndpoint;
+  if (networkParam) {
+    let endpointParam = ENDPOINTS.find(({ name }) => name === networkParam);
+    if (endpointParam) {
+      maybeEndpoint = endpointParam;
+    }
+  }
 
-  const [savedEndpoint, setEndpoint] = useLocalStorageState(
-    'connectionEndpoint',
-    ENDPOINTS[0].endpoint,
-  );
-  const endpoint = queryEndpoint || savedEndpoint;
+  if (networkStorage && !maybeEndpoint) {
+    let endpointStorage = ENDPOINTS.find(({ name }) => name === networkStorage);
+    if (endpointStorage) {
+      maybeEndpoint = endpointStorage;
+    }
+  }
 
-  const connection = useMemo(
-    () => new Connection(endpoint, 'recent'),
-    [endpoint],
-  );
+  const endpoint = maybeEndpoint || DEFAULT_ENDPOINT;
+
+  const { current: connection } = useRef(new Connection(endpoint.url));
 
-  const env =
-    ENDPOINTS.find(end => end.endpoint === endpoint)?.name || ENDPOINTS[0].name;
+  const [tokens, setTokens] = useState<Map<string, TokenInfo>>(new Map());
 
-  const [tokens, setTokens] = useState<TokenInfo[]>([]);
-  const [tokenMap, setTokenMap] = useState<Map<string, TokenInfo>>(new Map());
   useEffect(() => {
-    // fetch token files
```

Comment:
On master branch, changing the "network" (or "env" as it's called there), while this token fetching is ongoing would cause undefined behavior.